middleware/interpolate.js

```diff
@@ -21,132 +20,96 @@ example response from interpolation web service:
 
-    // perform interpolations asynchronously for all relevant hits
-    var timer = (new Date()).getTime();
-    async.map( res.data, bound, function( err, results ){
+    // only interpolate the street-layer results
+    // save this off into a separate array so that when docs are annotated
+    //  after the interpolate results are returned, no complicated bookkeeping is needed
+    const street_results = _.get(res, 'data', []).filter(result => result.layer === 'street');
 
-      // update res.data with the mapped values
-      if( !err ){
-        res.data = results;
+    // perform interpolations asynchronously for all relevant hits
+    const start = (new Date()).getTime();
+    async.map(street_results, req_bound_service, (err, interpolation_results) => {
+      if (err) {
+        logger.error(`[middleware:interpolation] ${_.defaultTo(err.message, err)}`);
+        return next();
       }
 
-      // sort the results to ensure that addresses show up higher than street centroids      
-      res.data = res.data.sort((a, b) => {
-        if (a.layer === 'address' && b.layer !== 'address') { return -1; }
-        if (a.layer !== 'address' && b.layer === 'address') { return 1; }
-        return 0;
-      });
-
-      // log the execution time, continue
-      logger.info( '[interpolation] [took]', (new Date()).getTime() - timer, 'ms' );
-      next();
-    });
-  };
-
-  middleware.transport = transport;
-  return middleware;
-}
-
-function interpolate( parsed_text, hit, cb ){
-
-  // no-op, this hit is not from the 'street' layer
-  // note: no network request is performed.
-  if( !hit || hit.layer !== 'street' ){
-    return cb( null, hit );
-  }
-
-  // query variables
-  var coord = hit.center_point;
-  var number = parsed_text.number;
-  var street = hit.address_parts.street || parsed_text.street;
-
-  // query interpolation service
-  this.query( coord, number, street, function( err, data ){
+      interpolation_results.forEach((interpolation_result, idx) => {
+        const source_result = street_results[idx];
+
+        // invalid / not useful response, debug log for posterity
+        // note: leave this hit unmodified
+        if (!_.has(interpolation_result, 'properties')) {
+          logger.debug(`[interpolation] [miss] ${req.clean.parsed_text}`);
+          return;
+        }
+
+        // the interpolation service returned a valid result, debug log for posterity
+        // note: we now merge those values with the existing 'street' record
+        logger.debug(`[interpolation] [hit] ${req.clean.parsed_text} ${JSON.stringify(interpolation_result)}`);
+
+        // -- metatdata --
+        source_result.layer = 'address';
+        source_result.match_type = 'interpolated';
+
+        // -- name --
+        source_result.name.default = `${interpolation_result.properties.number} ${source_result.name.default}`;
+
+        // -- source --
+        if (interpolation_result.properties.source === 'OSM') {
```

Comment:
mappings between the various permutations of our source names exist in https://github.com/pelias/api/blob/master/helper/type_mapping.js#L26 to keep all that knowledge in one place. It appears to not have been updated in a long time, for example to support polylines, but it feels like we should use it here.